modules/library/data/src/main/java/org/geotools/data/memory/MemoryFeatureWriter.java

```diff
@@ -111,7 +115,15 @@ public void write() throws IOException {
         if (current == null) {
             throw new IOException("No feature available to write");
         }
-    
+        //[GEOT-5683] preserve FeatureIDd during add feature
+        if (Boolean.TRUE.equals(current.getUserData().get(Hints.USE_PROVIDED_FID))) {
+            if (current.getUserData().containsKey(Hints.PROVIDED_FID)) {
+                String fid = (String) current.getUserData().get(Hints.PROVIDED_FID);
+                FeatureId id = new FeatureIdImpl(fid);
+                current = new SimpleFeatureImpl(current.getAttributes(), current.getFeatureType(), id);
+            }
+        }
```

Comment:
The MemoryDataStoreTest contains two tests that provide coverage of this functionality, _testGetFeatureStoreTransactionSupport_ and _testGetFeatureWriterTransaction_. These tests use a DefaultTransaction that sets the **USE_PROVIDED_FID** flag to true. I can add another test that sets this flag manually if necessary